### PR TITLE
[ramda] make coverage compose and pipe more strict

### PR DESCRIPTION
**Pull request by adispring**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/ramda/index.d.ts [hunk 1]
```diff
@@ -400,7 +401,16 @@ export function contains<T>(a: T): (list: readonly T[]) => boolean;
  * function is applied to those same arguments. The results of each branching function
  * are passed as arguments to the converging function to produce the return value.
  */
-export function converge<TArgs extends readonly any[], TResult>(after: ((...args: readonly any[]) => TResult), fns: Array<((...args: TArgs) => any)>): (...args: TArgs) => TResult;
+// tslint:disable:max-line-length
+export function converge<TArgs extends any[], TResult, R1, R2, R3, R4, R5, R6, R7, RestFunctions extends Array<(...args: TArgs) => any>>(converging: ((...args: readonly [R1, R2, R3, R4, R5, R6, R7, ...ReturnTypesOfFns<RestFunctions>]) => TResult), branches: [((...args: TArgs) => R1), ((...args: TArgs) => R2), ((...args: TArgs) => R3), ((...args: TArgs) => R4), ((...args: TArgs) => R5), ((...args: TArgs) => R6), ((...args: TArgs) => R7), ...RestFunctions]): (...args: TArgs) => TResult;
```
Comment (Nicholaiii) [inline types/ramda/index.d.ts:405]: This is getting painfully long, we should probably re-enable the tslint rule and break up the line, for easier understanding and maintainability.

**Review by Nicholaiii — COMMENTED** [review]
I really like this approach more than the previous fallback. However, I do believe we should start splitting up the functions into multiple lines. There are also some TODOs, is this WIP or do you want to merge now?

**Comment by adispring** [comment]
> I really like this approach more than the previous fallback. However, I do believe we should start splitting up the functions into multiple lines. There are also some TODOs, is this WIP or do you want to merge now?

I think we can merge it first. These todos are too hard to solve recently.

File: types/ramda/index.d.ts (hunk 1 above)
Comment (adispring) [reply types/ramda/index.d.ts:405]: Multiple lines style is ugly too, IMO.
Do you have any idea how to make multi lines style less ugly and more readable.

File: types/ramda/index.d.ts [hunk 2]
```diff
@@ -293,15 +294,34 @@ export function complement<TArgs extends any[]>(pred: (...args: TArgs) => unknow
  * functions must be unary.
  */
 // tslint:disable:max-line-length
+export function compose<TArgs extends any[], R1, R2, R3, R4, R5, R6, R7, TResult>(
```
Comment (adispring) [inline types/ramda/index.d.ts:297]: @Nicholaiii  I try to change it to multiple lines. what's your opinions of this style?

File: types/ramda/index.d.ts (hunk 1 above)
Comment (Nicholaiii) [reply types/ramda/index.d.ts:415]: IMHO this is improves readability and maintainability a lot:

```ts
export function converge<
  TArgs extends any[],
  TResult,
  R1, R2, R3, R4, R5, R6, R7,
  RestFunctions extends Array<(...args: TArgs) => any>
> (
    converging: (...args: readonly [
        R1, R2, R3, R4, R5, R6, R7,
        ...ReturnTypesOfFns<RestFunctions>
    ]) => TResult,
    branches: [
        ((...args: TArgs) => R1),
        ((...args: TArgs) => R2),
        ((...args: TArgs) => R3),
        ((...args: TArgs) => R4),
        ((...args: TArgs) => R5),
        ((...args: TArgs) => R6),
        ((...args: TArgs) => R7),
        ...RestFunctions
    ]
): (...args: TArgs) => TResult;
```

I agree that it's ugly, but it's not going to be pretty any way.

Alternatively for converging's args we could do:

```ts
    converging: (
        ...args: readonly [
            R1, R2, R3, R4, R5, R6, R7,
            ...ReturnTypesOfFns<RestFunctions>
        ]
    ) => TResult,
```

I don't think it adds anymore readability but I guess some people would call it "more correct". I like the leaner, first example

File: types/ramda/index.d.ts (hunk 2 above)
Comment (Nicholaiii) [reply types/ramda/index.d.ts:288]: I generally prefer 2-space indentation like you use, but the rest of the document uses four spaces. I also think putting the closing bracket and final return statement on same level as export, but you can see this in my example

File: types/ramda/index.d.ts (hunk 1 above)
Comment (adispring) [reply types/ramda/index.d.ts:415]: ok, I will modify the functions which length violate tslint.

File: types/ramda/index.d.ts (hunk 2 above)
Comment (Nicholaiii) [reply types/ramda/index.d.ts:288]: I think we should go for 4, for uniformity.

Comment (adispring) [reply types/ramda/index.d.ts:288]: > 

OK

Comment (adispring) [reply types/ramda/index.d.ts:287]: Done

**Review by Nicholaiii — APPROVED** [review]

**Comment by adispring** [comment]
@Nicholaiii  Thanks

**Pull request merged**